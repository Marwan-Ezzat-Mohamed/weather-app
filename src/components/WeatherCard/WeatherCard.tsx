import React from "react";
import { useNavigate } from "react-router-dom";
import { AiOutlineClose } from "react-icons/ai";
import WeatherIcon from "@components/WeatherIcon";
import { City } from "@/types";
import { formatDate } from "@/utils";

interface WeatherCardProps {
  city: City;
  temp: number;
  localTime: string;
  weatherConditionText: string;
  isDay: boolean;
  id: string;
  onRemoveClick: (id: string) => void;
}

const WeatherCard: React.FC<WeatherCardProps> = ({
  city,
  temp,
  localTime,
  weatherConditionText,
  isDay,
  id,
  onRemoveClick,
}) => {
  const navigate = useNavigate();

  console.log({localTime})
  const formattedDate = formatDate(localTime);

  const handleIconClick = (e: React.MouseEvent) => {
    e.stopPropagation();
    onRemoveClick(id);
  };

  const handleCardClick = () => {
    navigate({
      pathname: "/city",
      search: `?lat=${city.lat}&lng=${city.lng}`,
    });
  };

  return (
    <div
      onClick={handleCardClick}
      className="min-w-sm bg-secondary 0 relative flex w-full max-w-sm cursor-pointer justify-center rounded-2xl p-5 transition duration-200 hover:scale-105 hover:shadow-2xl "
    >
      <div className="w-full space-y-5">
        <div className="flex justify-between">
          <span className="rounded text-xl font-semibold">{city.name}</span>
          <div
            className="flex gap-2"
            onClick={handleIconClick}
            data-testid="remove-icon"
          >
            <AiOutlineClose className="text-3xl text-red-500 hover:text-red-700" />
          </div>
        </div>

        <div className="flex items-center justify-between">
          <div className="flex items-center justify-center">
            <WeatherIcon
              weatherConditionText={weatherConditionText}
              isDay={isDay}
            />
            <h1 className="text-6xl">{temp.toFixed(0)}Â°C</h1>
          </div>
          <div className="flex items-center justify-center">
            <span className="text-xl font-bold">{formattedDate}</span>
          </div>
        </div>
      </div>
    </div>
  );
};

export default WeatherCard;
