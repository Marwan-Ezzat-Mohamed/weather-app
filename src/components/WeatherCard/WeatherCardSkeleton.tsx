const WeatherCardSkeleton = () => (
  <div className="min-w-sm relative flex w-full max-w-sm animate-pulse justify-center rounded-2xl bg-gray-300 p-5 dark:bg-gray-600">
    <div className="w-full space-y-5">
      <div className="flex justify-between">
        <span className="h-7 w-2/3 animate-pulse rounded bg-gray-400 dark:bg-gray-500"></span>
        <div className="flex gap-2">
          <div className="h-7 w-7 animate-pulse text-3xl text-gray-400"></div>
        </div>
      </div>

      <div className="flex items-center justify-between">
        <div className="flex items-center justify-center gap-2">
          <div className="h-24 w-24 animate-pulse rounded-full bg-gray-400 dark:bg-gray-500"></div>
          <span className="h-20 w-24 animate-pulse rounded bg-gray-400 text-6xl dark:bg-gray-500"></span>
        </div>
        <div className="flex items-center justify-center">
          <span className="h-7 w-20 animate-pulse rounded bg-gray-400 text-xl font-bold dark:bg-gray-500"></span>
        </div>
      </div>
    </div>
  </div>
);

export { WeatherCardSkeleton };
