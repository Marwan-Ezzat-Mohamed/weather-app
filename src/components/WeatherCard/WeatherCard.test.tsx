import { render, fireEvent, screen } from "@testing-library/react";
import WeatherCard from "./WeatherCard";
import { MemoryRouter as Router } from "react-router-dom";

const mockUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockUsedNavigate,
}));

describe("WeatherCard Component", () => {
  const sampleCity = {
    id: "1",
    name: "Sample City",
    lat: 123.45,
    lng: 67.89,
  };

  const sampleProps = {
    city: sampleCity,
    temp: 25,
    localTime: "2023-09-20T15:30:00Z",
    weatherConditionText: "Sunny",
    isDay: true,
    id: "1",
    onRemoveClick: jest.fn(),
  };

  beforeEach(() => {
    sampleProps.onRemoveClick.mockClear();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Router>
        <WeatherCard {...sampleProps} />
      </Router>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the component with provided data", () => {
    render(
      <Router>
        <WeatherCard {...sampleProps} />
      </Router>
    );

    // Ensure that the component renders with correct city name and temperature
    expect(screen.getByText(sampleCity.name)).toBeInTheDocument();
    expect(screen.getByText(`${sampleProps.temp}Â°C`)).toBeInTheDocument();
  });

  it("handles card click", () => {
    render(<WeatherCard {...sampleProps} />);

    // Simulate a card click
    fireEvent.click(screen.getByText(sampleCity.name));

    // Ensure that navigate function was called with the correct pathname and search
    expect(mockUsedNavigate).toHaveBeenCalledWith({
      pathname: "/city",
      search: `?lat=${sampleCity.lat}&lng=${sampleCity.lng}`,
    });
  });

  it("handles remove icon click", () => {
    render(
      <Router>
        <WeatherCard {...sampleProps} />
      </Router>
    );

    // Simulate a remove icon click
    fireEvent.click(screen.getByTestId("remove-icon"));

    // Ensure that onRemoveClick prop was called with the correct id
    expect(sampleProps.onRemoveClick).toHaveBeenCalledWith("1");
  });
});
