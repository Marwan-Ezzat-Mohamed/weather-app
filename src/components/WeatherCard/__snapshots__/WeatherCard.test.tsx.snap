// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherCard Component matches snapshot 1`] = `
<div>
  <div
    class="min-w-sm relative flex w-full max-w-sm cursor-pointer justify-center rounded-2xl bg-slate-100 p-5 transition duration-200 hover:scale-105 hover:shadow-2xl dark:bg-slate-800"
  >
    <div
      class="w-full space-y-5"
    >
      <div
        class="flex justify-between"
      >
        <span
          class="rounded text-xl font-semibold"
        >
          Sample City
        </span>
        <div
          class="flex gap-2"
          data-testid="remove-icon"
        >
          <svg
            class="text-3xl text-red-500 hover:text-red-700"
            fill="currentColor"
            fill-rule="evenodd"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M799.855 166.312c.023.007.043.018.084.059l57.69 57.69c.041.041.052.06.059.084a.118.118 0 0 1 0 .069c-.007.023-.018.042-.059.083L569.926 512l287.703 287.703c.041.04.052.06.059.083a.118.118 0 0 1 0 .07c-.007.022-.018.042-.059.083l-57.69 57.69c-.041.041-.06.052-.084.059a.118.118 0 0 1-.069 0c-.023-.007-.042-.018-.083-.059L512 569.926 224.297 857.629c-.04.041-.06.052-.083.059a.118.118 0 0 1-.07 0c-.022-.007-.042-.018-.083-.059l-57.69-57.69c-.041-.041-.052-.06-.059-.084a.118.118 0 0 1 0-.069c.007-.023.018-.042.059-.083L454.073 512 166.371 224.297c-.041-.04-.052-.06-.059-.083a.118.118 0 0 1 0-.07c.007-.022.018-.042.059-.083l57.69-57.69c.041-.041.06-.052.084-.059a.118.118 0 0 1 .069 0c.023.007.042.018.083.059L512 454.073l287.703-287.702c.04-.041.06-.052.083-.059a.118.118 0 0 1 .07 0Z"
            />
          </svg>
        </div>
      </div>
      <div
        class="flex items-center justify-between"
      >
        <div
          class="flex items-center justify-center"
        >
          <canvas
            height="70"
            width="70"
          />
          <h1
            class="text-6xl"
          >
            25
            Â°C
          </h1>
        </div>
        <div
          class="flex items-center justify-center"
        >
          <span
            class="text-xl font-bold"
          >
            Wed 5:30 PM
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
